crates/hir-def/src/path/lower.rs

```diff
@@ -17,23 +17,58 @@ use crate::{
     type_ref::{LifetimeRef, TypeBound, TypeRef},
 };
 
+#[cfg(test)]
+thread_local! {
+    /// This is used to test `hir_segment_to_ast_segment()`. It's a hack, but it makes testing much easier.
+    pub(super) static SEGMENT_LOWERING_MAP: std::cell::RefCell<rustc_hash::FxHashMap<ast::PathSegment, usize>> = std::cell::RefCell::default();
+}
+
 /// Converts an `ast::Path` to `Path`. Works with use trees.
 /// It correctly handles `$crate` based path from macro call.
+// If you modify the logic of the lowering, make sure to check if `hir_segment_to_ast_segment()`
+// also needs an update.
 pub(super) fn lower_path(ctx: &mut LowerCtx<'_>, mut path: ast::Path) -> Option<Path> {
     let mut kind = PathKind::Plain;
     let mut type_anchor = None;
     let mut segments = Vec::new();
     let mut generic_args = Vec::new();
+    #[cfg(test)]
+    let mut ast_segments = Vec::new();
+    #[cfg(test)]
+    let mut ast_segments_offset = 0;
+    #[allow(unused_mut)]
+    let mut push_segment = |_segment: &ast::PathSegment, segments: &mut Vec<Name>, name| {
+        #[cfg(test)]
+        ast_segments.push(_segment.clone());
+        segments.push(name);
+    };
     loop {
         let segment = path.segment()?;
 
-        if segment.coloncolon_token().is_some() {
-            kind = PathKind::Abs;
+        if let Some(coloncolon) = segment.coloncolon_token() {
+            if coloncolon
+                .prev_token()
+                .is_none_or(|it| !path.syntax().text_range().contains_range(it.text_range()))
+            {
+                // A `::` inside segment (not between segments) can occur in two cases:
+                //  1. At the prefix of the path, i.e. `::foo::bar`. In this case the path is absolute.
+                //  2. In a truly bizarre syntax `Fn::(A, B) -> C`, which is another way to spell `Fn(A, B) -> C`.
```

Comment:
That's a valid parse? 🤨 That almost seems accidental

Reply:
Nevermind, the args are treated as generic arguments but with `()` so that makes sense. Our parse feels wrong though, I'll fix that